# Базовые структуры данных


## Индексация списка

**Цель блока заданий:** в данном блоке задач вы потренируетесь вызывать элементы или группы элементов из списка. 

**Рекомендации к выполнению блока заданий:** попробуйте разделить задание на части, поэтапно вызывайте по индексу составные части (которые могут также являться списками) из списка, пока не дойдёте до нужного элемента.

**Критерии оценки:** задания считаются выполненными, если выдача строго соответствует примеру из каждого задания. Допустим, в пятом задании результатом должен быть список с единственным элементом, равным 74.

In [ ]:
L = [[[1, 2, 3, 4, 5, 6, 7, 8, 9, 10], [11, 12, 13, 14, 15, 16, 17, 18, 19, 20]], 
    [[21, 22, 23, 24, 25, 26, 27, 28, 29, 30], [31, 32, 33, 34, 35, 36, 37, 38, 39, 40]], 
    [[41, 42, 43, 44, 45], [46, [47, 48], 49, 50], [51, 52, 53, 54, 55], [56, 57, 58, 59, 60]], 
    [61, 62, 63, [64, 65, 66, 67, 68, 69, 70, 71], 72, 73, 74, [75, [76, 77, 78], 79], 80], 
    [81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100]]

### 1.
Из приведённого выше списка списков выведите с помощью индексов число `7`.

In [ ]:
### YOUR CODE HERE ###
L[0][0][6]

7

### 2.
Из приведённого выше списка списков выведите с помощью индексов число `49`.

In [ ]:
### YOUR CODE HERE ###
L[2][1][0]

46

### 3.
Из приведённого выше списка списков выведите с помощью индексов число `78`.

In [ ]:
### YOUR CODE HERE ###
L[3][7][1][2]

78

### 4.
Из приведённого выше списка списков выведите с помощью индексов (используя **отрицательные значения**) число `99`.

In [ ]:
### YOUR CODE HERE ###

L[-1][-2]

99

### 5.
Из приведённого выше списка списков сделайте **срез** с числом `74`. 

In [ ]:
### YOUR CODE HERE ###

L[3][6:7]

[74]

### 6.
Из приведённого выше списка списков сделайте **срез** с числами `47, 48`. 

In [ ]:
### YOUR CODE HERE ###

l_1 = L[2][1][1:2]
l_2 = L[2][1][1]
print(l_1,"or", l_2)

[[47, 48]] or [47, 48]


### 7.
Из приведённого выше списка списков сделайте **срез** с числами `77, 78`. 

In [ ]:
### YOUR CODE HERE ###

L[3][7][1][1:]

[77, 78]

## Функция range()

**Цель блока заданий:** в данном блоке задач вы закрепите навык генерации числовых последовательностей через функцию range(). В дальнейшем функция range() будет часто использоваться в счётных циклах, а также для решения других задач.

**Рекомендации к выполнению блока заданий:** обратите внимание, что шаг последовательности может быть как «вперед», так и «назад». Для вывода результата в формате списка можно использовать функцию list().

**Критерии оценки:** задания считаются выполненными, если выдача строго соответствует примеру из задания. Использование функций сортировки не допускается.

### 8.
С помощью функции `range` сгенерируйте следующую последовательность: <br>
`[0, 1, 2, 3, 4]`

In [ ]:
### YOUR CODE HERE ###
list(range(5))


[0, 1, 2, 3, 4]

### 9.
С помощью функции `range` сгенерируйте следующую последовательность: <br>
`[3, 4]`

In [ ]:
### YOUR CODE HERE ###
list(range(3,5))


[3, 4]

### 10.
С помощью функции `range` сгенерируйте следующую последовательность: <br>
`[4, 8, 12, 16]`

In [ ]:
### YOUR CODE HERE ###
list(range(4,17,4))


[4, 8, 12, 16]

### 11.
С помощью функции `range` сгенерируйте следующую последовательность: <br>
`[16, 12, 8, 4]`

In [ ]:
### YOUR CODE HERE ###
list(range(16,3,-4))


[16, 12, 8, 4]

### 12.
С помощью функции `range` сгенерируйте следующую последовательность: <br>
`[0, 5, 10]`

In [ ]:
### YOUR CODE HERE ###
list(range(0,11,5))


[0, 5, 10]

### 13.
С помощью функции `range` сгенерируйте следующую последовательность: <br>
`[10, 5, 0]`

In [ ]:
### YOUR CODE HERE ###
list(range(10,-1,-5))


[10, 5, 0]

### 14.
С помощью функции `range` сгенерируйте следующую последовательность: <br>
`[-10, -9, -8, -7, -6, -5, -4, -3, -2, -1]`

In [ ]:
### YOUR CODE HERE ###
list(range(-10,0,1))


[-10, -9, -8, -7, -6, -5, -4, -3, -2, -1]

### 15.
С помощью функции `range` сгенерируйте следующую последовательность: <br>
`[-10, -16, -22, -28, -34]`

In [ ]:
### YOUR CODE HERE ###

list(range(-10,-35,-6))

[-10, -16, -22, -28, -34]

## Операторы и типы данных

Этот блок должен быть решён **БЕЗ** использования готовых функций и сторонних библиотек. Используйте операторы.

**Цель блока заданий:** в данном блоке задач вы потренируетесь в использовании базовых операторов в Python.

**Рекомендации к выполнению блока заданий:** выполнить данные задания можно опираясь на знания, полученные в этом и предшествующих модулях. Подумайте, как можно решить задачу простыми инструментами, если вам кажется, что для её выполнения необходим метод, с которым вы не знакомились в этом и предшествующих модулях.

**Критерии оценки:** задания считаются выполенными, если программа выводит корректный ответ в нужном формате. Формат выдачи приведён к каждому заданию. В решении не должны быть задействованы циклы, условия, функции, исключения, которые являются темами следующих модулей. Проверка корректности входных значений не требуется.

### 16.
Часы показывают время в формате h:mm:ss (количество часов (от 0 до 23), двузначное количество минут, двузначное количество секунд). Количество минут и секунд при необходимости дополняются до двузначного числа нулями.

Программа должна запрашивать количество секунд S, которое прошло с начала суток, и выводить в формате h:mm:ss, какое время будут показывать часы. <br><br>

*Формат ввода:* <br>
3600<br><br>

*Формат вывода:* <br>
1:00:00

In [ ]:
### YOUR CODE HERE ###
S = int(input("Enter count of seconds:"))
h = S//3600
dm=(S-h*3600)//600
m=(S-h*3600-dm*600)//60
sm=(S%60)//10
s=(S%60)%10
print(f"{h}:{dm}{m}:{sm}{s}")

Enter count of seconds:80457
22:20:57


### 17.
Машина выезжает на нулевой километр МКАД и едет по часовой стрелке с постоянной скоростью V километров в час. На каком километре МКАД машина окажется через T часов? 
<br> Длина МКАД: 109 км. <br><br> 

*Формат ввода:* <br>
60 <br>
1
<br><br>
*Формат вывода:* <br>
60

In [ ]:
### YOUR CODE HERE ###
V = int(input("Enter speed: "))
T = int(input("Enter time: "))
S_0 = 109
S = V * T
S_otn = S%S_0 
print(S_otn)

Enter speed: 120
Enter time: 1
11


### 18.
Напишите программу, которая запрашивает целое пятизначное число и выводит сумму его цифр.

*Формат ввода:* <br>
11111
<br><br>
*Формат вывода:* <br>
5

In [ ]:
### YOUR CODE HERE ###
d = input("Enter digit: ")
A = int(d[0]) + int(d[1]) + int(d[2]) + int(d[3]) + int(d[4])
print(A)


Enter digit: 99999
45


### 19.
Напишите программу, которая запрашивает целое четырёхзначное число и меняет местами его две первые и две последние цифры.

*Формат ввода:* <br>
5236 
<br><br>
*Формат вывода:* <br>
3652

In [ ]:
### YOUR CODE HERE ###
D_0 = input("Enter digit:")
D_1 = D_0.replace(D_0[0] + D_0[1] + D_0[2] + D_0[3], D_0[1] + D_0[0] + D_0[3] + D_0[2])
print(D_1)

Enter digit:8291
2819


### 20.
Напишите программу, которая на основании параметра D (сколько километров в день преодолевает машина) рассчитывает, сколько дней она затратит на путь длиной P километров.  <br><br>
*Формат ввода:* <br>
105 <br>
120
<br><br>
*Формат вывода:* <br>
2

In [ ]:
### YOUR CODE HERE ###
D = int(input("Enter car path: "))
P = int(input("Enter path: "))
A = P // D
b = P%D
A = A + int(b > 0)
print(A)


Enter car path: 105
Enter path: 210
2


### 21.
Напишите программу, которая считает, сколько рублей и копеек нужно заплатить за N авокадо, если одно авокадо стоит R рублей и K копеек. <br><br>
*Формат ввода:* <br>
40 20 <br>
5
<br><br>
*Формат вывода:* <br>
201 руб. 00 коп.

In [ ]:
### YOUR CODE HERE ###
R = int(input("Rubles: "))
K = int(input("Kopeiki: "))
N = int(input("Need avocado: "))
SR = R*N + (K*N)//100
SKD = ((K*N)%100)//10
SK = ((K*N)%100)%10
I = f"{SR} руб. {SKD}{SK} коп."
print(I)

Rubles: 21
Kopeiki: 95
Need avocado: 19
417 руб. 05 коп.


### 22.
Пусть организаторы мероприятия неправильно составили гугл-форму и просили людей указывать ФИО в неправильном порядке. Сначала спрашивали имя, потом отчество, затем фамилию. Напишите программу, которая будет переставлять ФИО в нужный порядок. <br> <br> 

*Формат ввода:* <br>
Иван Иванович Иванов<br><br>
*Формат вывода:* <br>
Иванов Иван Иванович

In [ ]:
### YOUR CODE HERE ###
SW=input("Enter your's name:")
D = SW.split(" ")
SR = D[2] + " " + D[0] + " " + D[1]
print(SR)


Enter your's name:Иван Иванович Иванов
Иванов Иван Иванович


## Cловари и списки

**Цель блока заданий:** в данном блоке задач вы зарепите навык работы со словарями и списками. В текущем блоке также пригодятся навыки, отработанные в первом блоке настоящего домашнего задания.

**Рекомендации к выполнению блока заданий:** попробуйте работать с большим словарём по тому же принципу, по которому вы вызывали составные элементы из большого списка в первом блоке заданий настоящей работы. Поэтапно вызывайте сначала один многосоставный объект, затем из него извлеките другой, но уже меньший, пока не доберётесь до целевого элемента.

**Критерии оценки:** задания считаются выполненными, если произошли нужные изменения (или их отстутсвие, если того требует условие задания), но при этом не произошло никаких других изменений, которые не описаны в задании. Обратите, пожалуйста, внимание, что программа в 23-й задаче должна быть универсальной, то есть работать не только для нулевого элемента.

### 23.
У вас есть список с заказом в ресторане. Напишите программу, которая меняет указанное по названию блюдо на другое. При этом новое блюдо в списке будет расположено на месте того, что было заменено.

*Формат ввода:* <br>
белое вино<br>
красное вино <br><br>
*Формат вывода:* <br>
`['красное вино', 'салат Цезарь', 'паста Карбонара', 'чизкейк', 'шоколадный сорбет']`

In [ ]:
order = ['белое вино', 'салат Цезарь', 'паста Карбонара', 'чизкейк', 'шоколадный сорбет']

In [ ]:
### YOUR CODE HERE ###
f1 = input("Enter what u want to del: ")
f2 = input("Enter what u want to insert: ")
a = order.index(f1)
order[a] = f2
print(order)



Enter what u want to del:чизкейк
Enter what u want to insert:пиво
['белое вино', 'салат Цезарь', 'паста Карбонара', 'пиво', 'шоколадный сорбет']


__________________
Вам дан словарь c расписанием на неделю. 

In [ ]:
diary = {'понедельник': {
                          'утро': ['погулять с собакой'],
                          'день': [],
                          'вечер': ['погулять с собакой']
                                      },
        'вторник': {
                          'утро': ['погулять с собакой'],
                          'день': [],
                          'вечер': ['погулять с собакой']
                                      },
        'среда': {
                          'утро': ['погулять с собакой'],
                          'день': [],
                          'вечер': ['погулять с собакой']
                                      },
        'четверг': {
                          'утро': ['погулять с собакой'],
                          'день': [],
                          'вечер': ['погулять с собакой']
                                      },
        'пятница': {
                          'утро': ['заехать в шиномонтаж', 'помыть машину'],
                          'день': [],
                          'вечер': ['поход в театр',  'ужин в кафе']
                                      },
        'суббота': {
                          'утро': [],
                          'день': [],
                          'вечер': []
                                      },
        'воскресенье': {
                          'утро': [],
                          'день': [],
                          'вечер': []
                                      }}

### 24.
Удалите ключи `суббота` и `воскресенье`. Вместо них добавьте пару, где ключ — это кортеж `суббота, воскресенье`, а значение — список с делом `посадить цветы на даче`. Можно ли сделать ключ `суббота, воскресенье` списком?

In [ ]:
### YOUR CODE HERE ###
diary.pop("суббота")
diary.pop("воскресенье")
diary.setdefault(("суббота","воскресенье"),["посадить цветы на даче"])

#нет, так как list изменяемый объект и не может быть задан в качестве ключа
#diary.setdefault(["суббота","воскресенье"],["посадить цветы на даче"])

print(diary)

{'понедельник': {'утро': ['погулять с собакой'], 'день': [], 'вечер': ['погулять с собакой']}, 'вторник': {'утро': ['погулять с собакой'], 'день': [], 'вечер': ['погулять с собакой']}, 'среда': {'утро': ['погулять с собакой'], 'день': [], 'вечер': ['погулять с собакой']}, 'четверг': {'утро': ['погулять с собакой'], 'день': [], 'вечер': ['погулять с собакой']}, 'пятница': {'утро': ['заехать в шиномонтаж', 'помыть машину'], 'день': [], 'вечер': ['поход в театр', 'ужин в кафе']}, ('суббота', 'воскресенье'): ['посадить цветы на даче']}


### 25.
Добавьте в список дел во вторник утром `поход к зубному`.

In [ ]:
### YOUR CODE HERE ###
diary["вторник"]["утро"].append("поход к зубному")
print(diary["вторник"])


{'утро': ['погулять с собакой', 'поход к зубному'], 'день': [], 'вечер': ['погулять с собакой']}


### 26.
Замените `поход в театр` на `поход в кино` в списке дел в пятницу вечером.

In [ ]:
### YOUR CODE HERE ###
i = diary["пятница"]["вечер"].index('поход в театр')
diary["пятница"]["вечер"][i] = "поход в кино"
print(diary["пятница"])


{'утро': ['заехать в шиномонтаж', 'помыть машину'], 'день': [], 'вечер': ['поход в кино', 'ужин в кафе']}


### 27.
Ваш друг вернётся из отпуска на один день раньше, поэтому он заберёт свою собаку в среду, а не в четверг. Удалите дело `погулять с собакой` из соответствующих списков.

In [ ]:
### YOUR CODE HERE ###
diary["четверг"]["утро"].remove('погулять с собакой')
diary["четверг"]["вечер"].remove('погулять с собакой')
print(diary["четверг"])

{'утро': [], 'день': [], 'вечер': []}


### 28.
Выведите второе дело из списка дел, которые вам нужно сделать в пятницу утром.

In [ ]:
### YOUR CODE HERE ###
diary["пятница"]["утро"][1]


'помыть машину'